service/content.go

```diff
@@ -129,16 +128,118 @@ func (s *NoneSerializer) ContentType() string {
+		data, err := ioutil.ReadAll(r)
+		if err != nil {
+			return err
+		}
+		*target = string(data)
+		return nil
+	case *[]byte:
+		data, err := ioutil.ReadAll(r)
+		if err != nil {
+			return err
+		}
+		*target = data
+		return nil
+	}
+	return invalidTypeForConsumer.Error(contentType, reflect.TypeOf(v))
+}
+
+// produce writes v to writer. v should be string, []byte, io.Reader.
+func (s *rawSerializer) produce(contentType string, w io.Writer, v interface{}) error {
+	if r, ok := v.(io.Reader); ok {
+		_, err := io.Copy(w, r)
+		return err
+	}
+	switch source := v.(type) {
+	case string:
+		_, err := io.WriteString(w, source)
+		return err
+	case []byte:
+		_, err := w.Write(source)
+		return err
+	}
+	return invalidTypeForProducer.Error(contentType, reflect.TypeOf(v))
+}
+
+// TextSerializer implements Consumer and Producer for content type "text/plain"
+type TextSerializer struct{ rawSerializer }
+
+// ContentType returns plain text MIME type.
+func (s *TextSerializer) ContentType() string {
+	return definition.MIMEText
+}
+
+// Consume reads data and converts it to string, []byte.
+func (s *TextSerializer) Consume(r io.Reader, v interface{}) error {
+	return s.consume(s.ContentType(), r, v)
+}
+
+// Produce writes v to writer. v should be string, []byte, io.Reader.
+func (s *TextSerializer) Produce(w io.Writer, v interface{}) error {
+	return s.produce(s.ContentType(), w, v)
+}
+
+// OctetStreamSerializer implements Consumer and Producer for content type "application/octet-stream"
+type OctetStreamSerializer struct{ rawSerializer }
+
+// ContentType returns octet stream MIME type.
+func (s *OctetStreamSerializer) ContentType() string {
+	return definition.MIMEOctetStream
+}
+
+// Consume reads data and converts it to string, []byte.
+func (s *OctetStreamSerializer) Consume(r io.Reader, v interface{}) error {
+	return s.consume(s.ContentType(), r, v)
+}
+
+// Produce writes v to writer. v should be string, []byte, io.Reader.
+func (s *OctetStreamSerializer) Produce(w io.Writer, v interface{}) error {
+	return s.produce(s.ContentType(), w, v)
+}
+
+// URLEncodedConsumer implements Consumer for content type "application/x-www-form-urlencoded"
+type URLEncodedConsumer struct{ rawSerializer }
+
+// ContentType returns url encoded MIME type.
+func (s *URLEncodedConsumer) ContentType() string {
+	return definition.MIMEURLEncoded
+}
+
+// Consume reads data and converts it to string, []byte.
+func (s *URLEncodedConsumer) Consume(r io.Reader, v interface{}) error {
+	return s.consume(s.ContentType(), r, v)
+}
+
+// FormDataConsumer implements Consumer for content type "multipart/form-data"
+type FormDataConsumer struct{ rawSerializer }
+
+// ContentType returns form data MIME type.
+func (s *FormDataConsumer) ContentType() string {
+	return definition.MIMEFormData
+}
+
+// Consume reads data and converts it to string, []byte.
+func (s *FormDataConsumer) Consume(r io.Reader, v interface{}) error {
+	return s.consume(s.ContentType(), r, v)
 }
```

Comment:
Just so easy for parsing?

Reply:
It works only if you don't register a `service.ParseRequestForm()` filter. If the filter is installed, the body of a `multipart/formdata` request would be parsed and you will get nothing by the consumer.

In most occasion, you should not use the consumer to get request body. `definition.File` and `definition.Form` is recommended.